### DISCUSSION TITLE
What is the correct way to handle user input to IAnsiConsole when implementing unit tests?

### DISCUSSION THREAD

**Discussion by wbaldoumas** [spectreconsole/spectre.console#1524] [Q&A] [answered]
I have some implementations that leverage the `IAnsiConsole.Prompt` , `IAnsiConsole.Ask`, and `IAnsiConsole.Confirm` methods. 

How can I mock user input for these within the context of unit tests? Since these are extension methods, there isn't an easy way to mock the behavior here and I didn't see an obvious way of using `TestConsole` to achieve this, but may have missed something there.

**Comment by mapersson** [comment]
Take a look at [TestConsoleInput](src/Spectre.Console.Testing/TestConsoleInput.cs), specifically ["PushText"](https://github.com/spectreconsole/spectre.console/blob/2cc6c457ade3c0b795ebb4d0f8868e1befc56a6c/src/Spectre.Console.Testing/TestConsoleInput.cs#L22) and ["PushTextWithEnter"](https://github.com/spectreconsole/spectre.console/blob/2cc6c457ade3c0b795ebb4d0f8868e1befc56a6c/src/Spectre.Console.Testing/TestConsoleInput.cs#L39).

I have used the following code in my unit test to mock user input:

`var console = new TestConsole();`
`console.Input.PushTextWithEnter("mock user input");`

**Comment by FrankRay78** [comment, marked as answer]
@wbaldoumas  There are quite a few interactive unit tests located here, which you might find provide the pattern you are looking for:  https://github.com/spectreconsole/spectre.console/tree/main/src/Tests/Spectre.Console.Tests/Unit/Prompts

Eg:

```csharp
    [Fact]
    public void Should_Select_The_First_Leaf_Item()
    {
        // Given
        var console = new TestConsole();
        console.Profile.Capabilities.Interactive = true;
        console.Input.PushKey(ConsoleKey.Enter);

        // When
        var prompt = new SelectionPrompt<string>()
                .Title("Select one")
                .Mode(SelectionMode.Leaf)
                .AddChoiceGroup("Group one", "A", "B")
                .AddChoiceGroup("Group two", "C", "D");
        var selection = prompt.Show(console);

        // Then
        selection.ShouldBe("A");
    }
```

**Reply by wbaldoumas** [reply]
Thanks, both! I think this will work for what I need. 👍
